### ISSUE TITLE
Make mathseminars.org more generic to replicate arXiv.org sections

### ISSUE THREAD

**Issue by yujitach** [roed314/seminars#46]
I’m a string theorist and I’m very much interested in adapting mathseminars.org for string theory. As arXiv.org is the de facto standard in both math and physics, it would be nice if mathseminars.org would be just seminars.org, and that each seminar can be tagged by one or more arXiv category tags, such as math.AT or hep-th. This hopefully minor enhancement would  instantly make mathseminars.org into an arXiv for web seminars.

**Comment by roed314** [comment]
We've thought about this, but seminars.org wasn't available as a domain name.  :-(

We decided to start with math, but are certainly open to this idea.  The missing ingredients are:
* An appropriate domain name (I wanted math.seminars.org and hep.seminars.org, etc but wasn't able to get seminars.org).  I think there's a discussion to be had about what the scope of the talks shown on the front page with no filtering would be.
* Some work on the code base, as you say, to update from assuming a "math." prefix on topics.
* Substantial interest from communities outside math in making this work.  In addition to writing the code, we've put a lot of effort into seeding the site with seminars and talks.  We would need people to take on that task for physics or other disciplines.

**Comment by bpioline** [comment]
Amazing work ! 

I totally agree with yujitach's suggestion. There is a lof of interest from the high energy theory community. Currently, lots of talks are advertized through a Facebook group "Physics in time of coronavirus" but this is hardly sustainable. The domain name "seminartracker.org" is up for grabs - at least now. Regarding the front page, it would be quite crowded on the first time, but less so on future visits if you apply by default previous, cookie-registered filters.

Alternatively (but less preferable imho), since people interested in both math and physics talks are likely to be a minority, you could open a second branch "physeminars.org" running the same code on a different database.

**Comment by jamiesully** [comment]
I'm currently managing the Facebook group for high-energy theory (https://www.facebook.com/groups/204251324219544) and it's clear there is substantial interest already. 

If the maintainers of mathseminars aren't interested in broadening outside of math for the time being, I'm happy to set-up a fork for the physics community.

**Comment by roed314** [comment]
I think we are willing to broaden (and, in particular, to host and maintain the database of talks), but need some more manpower from people who know enough Python/Flask/Django/Postgres to help resolve issues (and, of course, to administer other subject areas).  Right now we're already drowning a bit in terms of the features that we're trying to add immediately to make this work better for the math community.

I'm happy to have a Zoom meeting to discuss what would be required for either of these options.

**Comment by jamiesully** [comment]
I'm happy to help as I'm able. I have some minimal experience with Python and Flask, but definitely not an expert.

**Comment by roed314** [comment]
We just purchased scimeet.org.  Later tonight I'll think a bit about what needs to be done to make it usable by other subjects.

**Comment by julioparra** [comment]
Much needed! Trying to mimic the time tested arXiv model (categories, moderators, endorsement, daily announcement, etc) seems like a good idea. I have worked with Python, but not the other tools.  Still, I'd be happy to contribute if I can. For instance by curating the high-energy categories (hep-th, hep-ph, gr-qc), etc.

**Comment by jamiesully** [comment]
I also can offer up 'seminarxiv.org' if that's a preferable domain name.

**Comment by roed314** [comment]
Thanks for the offer on `seminarxiv.org`.  We can decide later what the right domain is, once we have the functionality going.

Here are some of my thoughts on what needs to be done.

* Structural questions: are there multiple databases each with a webserver, or one webserver running per subject but all connecting to the same database, or just one webserver overall (which is routing all the different subdomains)?  I'm writing the below assuming a single webserver, but I'm not convinced that's the best way to go.  In particular, if we use multiple databases then it would be possible to set up a completely separate physics site now (you guys would just run it on your own server and fork our code).  I think that this would be more work in the long run, and would make it more difficult for us to collaborate on common features, but it's probably a bit faster to get up and running.

* Schema changes: the following tables need to have `subject` column added with default "math" (and the current rows updated): `seminars`, `talks`, `topics`, `seminar_subscriptions`, `talk_subscriptions`, `seminar_organizers`.  The `users` table should probably change `creator` and `admin` columns from boolean to `text[]` columns, listing the subjects for which that user is a creator/admin.

* Update the user objects to use the subject (e.g. `is_admin(self, subject)`) and update the places that these functions are called.

* Update the contents of the ``topics`` table.  This will invalidate peoples' cookies, so provide an upgrade path in the javascript (detect topics without periods and rewrite them to include math.)  The update to the javascript code has to happen before the contents of the table are updated.

* Update all the routing functions to take the subject as an argument.  For example,
```python
@app.route("/", subdomain="<subject>")
def index(subject):
    pass
```

* Need to create the apache configuration files to route the dns requests to the right place (preserve the master.mathseminars.org that we just set up)

* Need to update all the relevant database interactions to include the subject

I'm not going to start on these immediately, but I would suggest that if someone wants to help they can follow the instructions in the `GettingStarted.md` document and try to set up a development environment locally.  You won't have the ability to change the database, but you can play around with the codebase.  I'm happy to create entirely separate tables (e.g. an `arxiv_topics` table that has a `subject` column).

We should probably also find a better way to collaborate than comments on a github issue.  If anyone wants to get started on this, send me an email (roed@mit.edu) and we can talk about better ways to work together.

**Comment by duetosymmetry** [comment]
Bumping this again during the APS April meeting. In the division of gravitational physics, somebody suggested a similar repository for talks in our specialty. I pointed out the existence of mathseminars.org and that the HEP community was also interested. So, here are some more votes for generalizing.

**Comment by roed314** [comment]
We made a lot of progress today.  I'm hopeful that we'll have something in beta by tomorrow.

**Comment by duetosymmetry** [comment]
Wonderful to hear, @roed314. I have web, python, and the smallest amount of database experience but no flask or django. Let me know if there's something I can do to help.

**Comment by roed314** [comment]
We're changing our approach, so I don't think this is going to get done today.  I'm hopeful that it will be done tomorrow though.  I'll make a comment shortly with the new plan.

**Comment by roed314** [comment]
Here's a todo list

- [ ] Add a welcome splash thing where users select subjects if they have no subject cookie.
- [ ] Add subject pages (e.g. scimeet.org/math) that ignores your subject cookie and just shows math talks.  You can still update your topics cookie which will persist on the front page.  Note that if you set some math topics but had physics subject filtering on the front page, the math topics in your topic cookie would have no effect (since you're already filtering them out).  But if you turned subject filtering off, your math topics would be relevant.
- [ ] Add topic pages (e.g. scimeet.org/math/nt) that ignore your subject and topics cookies.
- [ ] Change the topic selecter div to be tabbed by subject
- [ ] Add a filter by subject and a subject cookie so that users can say that they're just interested in math and physics and not all that other stuff.  There's another invisible div holding the subject selection that you can toggle visible/invisible independently of the topics div.
- [ ] Add an option to user info specifying which subjects they want to see topics from when creating seminars and talks.  This should initially populated from their browse subject cookie.  It requires a schema update
- [ ] Replace periods with underscores in abbreviations in topics table.  Add `math_` to the beginnings of the math topics.  Update code to deal with having topics from different subjects.
- [ ] Decide on a url (current leading candidates are `scimeet.org` and `seminarxiv.org`).  If we want to use `seminarxiv.org`, get permission from the `arxiv` people.
- [ ] Set up a site to view the `subject2` branch in action.  At first this will be using the development database.
- [ ] Once it looks good on the development database, make necessary schema changes to the live database and switch the testing site to use the live database.
- [ ] Figure out a transition plan for how to redirect visitors from `mathseminars.org` to the new url.
- [ ] Add a bunch of physics talks.

**Comment by roed314** [comment]
We're working on this branch: https://github.com/roed314/seminars/tree/subject2

**Comment by scratchmex** [comment]
I can help with the project in general. I have python experience in flask and backend.
